src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/DirectiveAttributeParameterCompletionItemProvider.cs

```diff
@@ -0,0 +1,147 @@
+            var owner = syntaxTree.Root.LocateOwner(change);
+
+            if (owner == null)
+            {
+                return Array.Empty<RazorCompletionItem>();
+            }
+
+            if (!TryGetAttributeInfo(owner, out var attributeName, out _, out var parameterName, out var parameterNameLocation))
+            {
+                // Either we're not in an attribute or the attribute is so malformed that we can't provide proper completions.
+                return Array.Empty<RazorCompletionItem>();
+            }
+
+            if (!parameterNameLocation.IntersectsWith(location.AbsoluteIndex))
+            {
+                // We're trying to retrieve completions on a portion of the name that is not supported (such as the name, i.e., |@bind|:format).
+                return Array.Empty<RazorCompletionItem>();
+            }
+
+            if (!TryGetElementInfo(owner.Parent.Parent, out var containingTagName, out var attributes))
+            {
+                // This should never be the case, it means that we're operating on an attribute that doesn't have a tag.
+                return Array.Empty<RazorCompletionItem>();
+            }
+
+            var completions = GetAttributeParameterCompletions(attributeName, parameterName, containingTagName, attributes, tagHelperDocumentContext);
+            return completions;
+        }
+
+        // Internal for testing
+        internal IReadOnlyList<RazorCompletionItem> GetAttributeParameterCompletions(
+            string attributeName,
+            string parameterName,
+            string containingTagName,
+            IEnumerable<string> attributes,
+            TagHelperDocumentContext tagHelperDocumentContext)
+        {
+            var descriptorsForTag = _tagHelperFactsService.GetTagHelpersGivenTag(tagHelperDocumentContext, containingTagName, parentTag: null);
+            if (descriptorsForTag.Count == 0)
+            {
+                // If the current tag has no possible descriptors then we can't have any additional attributes.
+                return Array.Empty<RazorCompletionItem>();
+            }
+
+            // Attribute parameters are case sensitive when matching
+            var attributeCompletions = new HashSet<string>(StringComparer.Ordinal);
+            foreach (var descriptor in descriptorsForTag)
+            {
+                for (var i = 0; i < descriptor.BoundAttributes.Count; i++)
+                {
+                    var attributeDescriptor = descriptor.BoundAttributes[i];
+                    var boundAttributeParameters = attributeDescriptor.BoundAttributeParameters;
+                    if (boundAttributeParameters.Count == 0)
+                    {
+                        continue;
+                    }
+
+                    if (TagHelperMatchingConventions.CanSatisfyBoundAttribute(attributeName, attributeDescriptor))
+                    {
+                        for (var j = boundAttributeParameters.Count - 1; j >= 0; j--)
```

Comment:
Why backwards?

Reply:
Worth a comment if there's a good reason because it doesn't seem super obvious.

Reply:
0 reason why it should be backwards lol